src/uct/ib/ud/base/ud_ep.c

```diff
@@ -684,6 +684,35 @@ static uct_ud_ep_t *uct_ud_ep_create_passive(uct_ud_iface_t *iface, uct_ud_ctl_h
     return ep;
 }
 
+static void uct_ud_ep_rx_ctl_drop_packet(uct_ud_ep_t *ep, uct_ud_neth_t *neth)
+{
+    uct_ud_ctl_hdr_t *ctl = (uct_ud_ctl_hdr_t*)(neth + 1);
+    int creq_received     = (ep->flags & UCT_UD_EP_FLAG_CREQ_RCVD) &&
+                            (ep->rx_creq_count > 1);
+    int crep_received     = ep->flags & UCT_UD_EP_FLAG_CREP_RCVD;
+
+    ucs_assert((ctl->type == UCT_UD_PACKET_CREQ) ||
+               (ctl->type == UCT_UD_PACKET_CREP));
+
+    ucs_trace_data("ep %p: drop %s with psn %u, head_sn %u",
+                   ep, (ctl->type == UCT_UD_PACKET_CREQ) ? "CREQ" : "CREP",
+                   neth->psn, ep->rx.ooo_pkts.head_sn);
+
+    ucs_assertv_always(/* Drop CREQ, if the endpoint received duplicate
+                        * CREQ */
+                       ((ctl->type == UCT_UD_PACKET_CREQ) && creq_received) ||
+                       /* Drop:
+                        * - CREP, if the endpoint received duplicate CREP
+                        * - CREQ, if the endpoint received CREQ after CREP */
+                       crep_received,
```

Comment:
when this function is called from uct_ud_ep_rx_ctl(), we already know it's crep. so no need assertion.
better move this assertion to line 770

Reply:
it checks that CREP has already been received. i.e. it is duplicate CREP

Reply:
maybe pass "exp_ep_flag" which could be either UCT_UD_EP_FLAG_CREQ_RCVD or UCT_UD_EP_FLAG_CREP_RCVD, and check (ep_flags & exp_ep_flag) ? it's not so important to check rx_creq_count>1

Reply:
good idea, done